core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala

```diff
@@ -187,47 +200,74 @@ private[history] class FsHistoryProvider(conf: SparkConf) extends ApplicationHis
         }
         .flatMap { entry =>
           try {
-            Some(replay(entry, new ReplayListenerBus()))
+            Some((getModificationTime(entry).get, entry))
           } catch {
             case e: Exception =>
               logError(s"Failed to load application log data from $entry.", e)
               None
           }
         }
-        .sortWith(compareAppInfo)
+        .sortWith(_._1 >= _._1)
+        .map { case (_, file) => file }
+
+      logInfos.sliding(20, 20).foreach { batch =>
+        replayExecutor.submit(new Runnable {
+          override def run(): Unit = mergeApplicationListing(batch)
+        })
+      }
 
       lastModifiedTime = newLastModifiedTime
+    } catch {
+      case e: Exception => logError("Exception in checking for event log updates", e)
+    }
+  }
 
-      // When there are new logs, merge the new list with the existing one, maintaining
-      // the expected ordering (descending end time). Maintaining the order is important
-      // to avoid having to sort the list every time there is a request for the log list.
-      if (!logInfos.isEmpty) {
-        val newApps = new mutable.LinkedHashMap[String, FsApplicationHistoryInfo]()
-        def addIfAbsent(info: FsApplicationHistoryInfo) = {
-          if (!newApps.contains(info.id) ||
-              newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
-              !info.logPath.endsWith(EventLoggingListener.IN_PROGRESS)) {
-            newApps += (info.id -> info)
-          }
-        }
+  /**
+   * Replay the log files in the list and merge the list of old applications with new ones
+   */
+  private def mergeApplicationListing(logs: Seq[FileStatus]): Unit = {
+    def addIfAbsent(
+        newApps: mutable.LinkedHashMap[String, FsApplicationHistoryInfo],
+        info: FsApplicationHistoryInfo): Unit = {
+      if (!newApps.contains(info.id) ||
+        newApps(info.id).logPath.endsWith(EventLoggingListener.IN_PROGRESS) &&
+          !info.logPath.endsWith(EventLoggingListener.IN_PROGRESS)) {
+        newApps += (info.id -> info)
+      }
+    }
 
-        val newIterator = logInfos.iterator.buffered
-        val oldIterator = applications.values.iterator.buffered
-        while (newIterator.hasNext && oldIterator.hasNext) {
-          if (compareAppInfo(newIterator.head, oldIterator.head)) {
-            addIfAbsent(newIterator.next)
-          } else {
-            addIfAbsent(oldIterator.next)
-          }
-        }
-        newIterator.foreach(addIfAbsent)
-        oldIterator.foreach(addIfAbsent)
+    def mergeApps(list: Seq[FsApplicationHistoryInfo]): Unit = {
```

Comment:
there's no reason for this to be in its own sub-method, since it's only called once at the end of `mergeApplicationListing`. I would just extract the content of this method out.